Non-sense help when type parameter from outer function is used

# Code

```rust
struct A;

impl A {
    fn banana(&mut self) {
        fn peach(this: &Self) {
            
        }
    }
}
```

# Error

```
error[E0401]: can't use type parameters from outer function
 --> src/main.rs:9:25
  |
7 | impl A {
  | ---- `Self` type implicitely declared here, on the `impl`
8 |     fn banana(&mut self) {
9 |         fn peach(this: &Self) {
  |            -----        ^^^^ use of type variable from outer function
  |            |
  |            help: try using a local type parameter instead: `peach<Self>`
```

---

It seems that the help message is trying to suggest to introduce a local type parameter, but is worded rather poorly, as if suggesting to *use* a `this: &peach<Self>`.

Perhaps rewording the help message to explicitly say "introduce" rather than "use" would be better, but even then `Self` is a keyword, and cannot be introduced as a local generic parameter anyway.

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -204,7 +204,7 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
                     }) {
                         err.span_label(
                             reduce_impl_span_to_impl_keyword(cm, impl_span),
-                            "`Self` type implicitly declared here, on the `impl`",
+                            "`Self` type implicitly declared here, by this `impl`",
                         );
                     }
                     match (maybe_trait_defid, maybe_impl_defid) {

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -197,15 +197,26 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
             err.span_label(span, "use of type variable from outer function");
 
             let cm = resolver.session.source_map();
-            let mut is_self = false;
             match outer_def {
-                Def::SelfTy(_, maybe_impl_defid) => {
-                    if let Some(impl_span) = maybe_impl_defid.map_or(None,
-                            |def_id| resolver.definitions.opt_span(def_id)) {
-                        err.span_label(reduce_impl_span_to_impl_keyword(cm, impl_span),
-                                    "`Self` type implicitly declared here, on the `impl`");
+                Def::SelfTy(maybe_trait_defid, maybe_impl_defid) => {
+                    if let Some(impl_span) = maybe_impl_defid.and_then(|def_id| {
+                        resolver.definitions.opt_span(def_id)
+                    }) {
+                        err.span_label(
+                            reduce_impl_span_to_impl_keyword(cm, impl_span),
+                            "`Self` type implicitly declared here, on the `impl`",
+                        );
                     }
-                    is_self = true;
+                    match (maybe_trait_defid, maybe_impl_defid) {
+                        (Some(_), None) => {
+                            err.span_label(span, "can't use `Self` here");
+                        }
+                        (_, Some(_)) => {
+                            err.span_label(span, "use a type here instead");
+                        }
+                        (None, None) => bug!("`impl` without trait nor type?"),
+                    }
+                    return err;
                 },
                 Def::TyParam(typaram_defid) => {
                     if let Some(typaram_span) = resolver.definitions.opt_span(typaram_defid) {
@@ -221,12 +232,7 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
             // Try to retrieve the span of the function signature and generate a new message with
             // a local type parameter
             let sugg_msg = "try using a local type parameter instead";
-            if is_self {
-                // Suggest using the actual type
-                err.span_label(span, "use a materialized type here instead");
-            } else if let Some(
-                (sugg_span, new_snippet),
-            ) = cm.generate_local_type_param_snippet(span) {
+            if let Some((sugg_span, new_snippet)) = cm.generate_local_type_param_snippet(span) {
                 // Suggest the modification to the user
                 err.span_suggestion_with_applicability(
                     sugg_span,

--- a/src/librustc_resolve/lib.rs
+++ b/src/librustc_resolve/lib.rs
@@ -197,6 +197,7 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
             err.span_label(span, "use of type variable from outer function");
 
             let cm = resolver.session.source_map();
+            let mut is_self = false;
             match outer_def {
                 Def::SelfTy(_, maybe_impl_defid) => {
                     if let Some(impl_span) = maybe_impl_defid.map_or(None,
@@ -204,6 +205,7 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
                         err.span_label(reduce_impl_span_to_impl_keyword(cm, impl_span),
                                     "`Self` type implicitly declared here, on the `impl`");
                     }
+                    is_self = true;
                 },
                 Def::TyParam(typaram_defid) => {
                     if let Some(typaram_span) = resolver.definitions.opt_span(typaram_defid) {
@@ -219,7 +221,12 @@ fn resolve_struct_error<'sess, 'a>(resolver: &'sess Resolver,
             // Try to retrieve the span of the function signature and generate a new message with
             // a local type parameter
             let sugg_msg = "try using a local type parameter instead";
-            if let Some((sugg_span, new_snippet)) = cm.generate_local_type_param_snippet(span) {
+            if is_self {
+                // Suggest using the actual type
+                err.span_label(span, "use a materialized type here instead");
+            } else if let Some(
+                (sugg_span, new_snippet),
+            ) = cm.generate_local_type_param_snippet(span) {
                 // Suggest the modification to the user
                 err.span_suggestion_with_applicability(
                     sugg_span,

--- a/src/test/ui/error-codes/E0401.stderr
+++ b/src/test/ui/error-codes/E0401.stderr
@@ -27,9 +27,10 @@ LL | impl<T> Iterator for A<T> {
    | ---- `Self` type implicitly declared here, on the `impl`
 ...
 LL |         fn helper(sel: &Self) -> u8 { //~ ERROR E0401
-   |            ------       ^^^^ use of type variable from outer function
-   |            |
-   |            help: try using a local type parameter instead: `helper<Self>`
+   |                         ^^^^
+   |                         |
+   |                         use of type variable from outer function
+   |                         use a materialized type here instead
 
 error: aborting due to 3 previous errors
 

--- a/src/test/ui/error-codes/E0401.stderr
+++ b/src/test/ui/error-codes/E0401.stderr
@@ -24,7 +24,7 @@ error[E0401]: can't use type parameters from outer function
   --> $DIR/E0401.rs:32:25
    |
 LL | impl<T> Iterator for A<T> {
-   | ---- `Self` type implicitly declared here, on the `impl`
+   | ---- `Self` type implicitly declared here, by this `impl`
 ...
 LL |         fn helper(sel: &Self) -> u8 { //~ ERROR E0401
    |                         ^^^^

--- a/src/test/ui/error-codes/E0401.stderr
+++ b/src/test/ui/error-codes/E0401.stderr
@@ -30,7 +30,7 @@ LL |         fn helper(sel: &Self) -> u8 { //~ ERROR E0401
    |                         ^^^^
    |                         |
    |                         use of type variable from outer function
-   |                         use a materialized type here instead
+   |                         use a type here instead
 
 error: aborting due to 3 previous errors
 

--- a/src/test/ui/issues/issue-12796.stderr
+++ b/src/test/ui/issues/issue-12796.stderr
@@ -2,9 +2,10 @@ error[E0401]: can't use type parameters from outer function
   --> $DIR/issue-12796.rs:13:22
    |
 LL |         fn inner(_: &Self) {
-   |            -----     ^^^^ use of type variable from outer function
-   |            |
-   |            help: try using a local type parameter instead: `inner<Self>`
+   |                      ^^^^
+   |                      |
+   |                      use of type variable from outer function
+   |                      use a materialized type here instead
 
 error: aborting due to previous error
 

--- a/src/test/ui/issues/issue-12796.stderr
+++ b/src/test/ui/issues/issue-12796.stderr
@@ -5,7 +5,7 @@ LL |         fn inner(_: &Self) {
    |                      ^^^^
    |                      |
    |                      use of type variable from outer function
-   |                      use a materialized type here instead
+   |                      can't use `Self` here
 
 error: aborting due to previous error
 

--- a/src/test/ui/use-self-in-inner-fn.rs
+++ b/src/test/ui/use-self-in-inner-fn.rs
@@ -1,0 +1,24 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct A;
+
+impl A {
+//~^ NOTE `Self` type implicitly declared here, on the `impl`
+    fn banana(&mut self) {
+        fn peach(this: &Self) {
+        //~^ ERROR can't use type parameters from outer function
+        //~| NOTE use of type variable from outer function
+        //~| NOTE use a materialized type here instead
+        }
+    }
+}
+
+fn main() {}

--- a/src/test/ui/use-self-in-inner-fn.rs
+++ b/src/test/ui/use-self-in-inner-fn.rs
@@ -11,7 +11,7 @@
 struct A;
 
 impl A {
-//~^ NOTE `Self` type implicitly declared here, on the `impl`
+//~^ NOTE `Self` type implicitly declared here, by this `impl`
     fn banana(&mut self) {
         fn peach(this: &Self) {
         //~^ ERROR can't use type parameters from outer function

--- a/src/test/ui/use-self-in-inner-fn.rs
+++ b/src/test/ui/use-self-in-inner-fn.rs
@@ -16,7 +16,7 @@ impl A {
         fn peach(this: &Self) {
         //~^ ERROR can't use type parameters from outer function
         //~| NOTE use of type variable from outer function
-        //~| NOTE use a materialized type here instead
+        //~| NOTE use a type here instead
         }
     }
 }

--- a/src/test/ui/use-self-in-inner-fn.stderr
+++ b/src/test/ui/use-self-in-inner-fn.stderr
@@ -2,7 +2,7 @@ error[E0401]: can't use type parameters from outer function
   --> $DIR/use-self-in-inner-fn.rs:16:25
    |
 LL | impl A {
-   | ---- `Self` type implicitly declared here, on the `impl`
+   | ---- `Self` type implicitly declared here, by this `impl`
 ...
 LL |         fn peach(this: &Self) {
    |                         ^^^^

--- a/src/test/ui/use-self-in-inner-fn.stderr
+++ b/src/test/ui/use-self-in-inner-fn.stderr
@@ -8,7 +8,7 @@ LL |         fn peach(this: &Self) {
    |                         ^^^^
    |                         |
    |                         use of type variable from outer function
-   |                         use a materialized type here instead
+   |                         use a type here instead
 
 error: aborting due to previous error
 

--- a/src/test/ui/use-self-in-inner-fn.stderr
+++ b/src/test/ui/use-self-in-inner-fn.stderr
@@ -1,0 +1,15 @@
+error[E0401]: can't use type parameters from outer function
+  --> $DIR/use-self-in-inner-fn.rs:16:25
+   |
+LL | impl A {
+   | ---- `Self` type implicitly declared here, on the `impl`
+...
+LL |         fn peach(this: &Self) {
+   |                         ^^^^
+   |                         |
+   |                         use of type variable from outer function
+   |                         use a materialized type here instead
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0401`.

